docs/5.x/writing-a-record-builder.md

```diff
@@ -0,0 +1,250 @@
+            Metrics::INDEX_SUM_VISIT_LENGTH => $row[Metrics::INDEX_SUM_VISIT_LENGTH],
+            Metrics::INDEX_BOUNCE_COUNT => $row[Metrics::INDEX_BOUNCE_COUNT],
+            Metrics::INDEX_NB_VISITS_CONVERTED => $row[Metrics::INDEX_NB_VISITS_CONVERTED],
+        ];
+
+        $report->sumRowWithLabel($row['label'] ?? '', $columns);
+    }
+
+    return [
+        'MyPlugin_myRecord' => $report,
+    ];
+}
+```
+
+This example queries the `log_visit` table, grouping by the `config_browser_name` column and aggregating visit metrics.
+Then, for each row of that query, it adds the metrics to a `DataTable` which is eventually returned.
+
+Most `aggregate()` methods will be more complicated than this, but hopefully it provides you with a general understanding
+of how they should work. We recommend looking at existing `RecordBuilder`s in Matomo as well to see what is possible.
+
+### Step four: decide whether you need to set custom row limits or aggregation operations
+
+At this point, the hard parts are over. The last two steps are just finishing touches.
+
+By default, Matomo does not limit the data that is inserted into archive tables. For reports that have a limited number
+of rows, like the `VisitorInterest.getVisitsByVisitCount` and `UserCountry.getCountry`, this is acceptable. But for reports
+with a variable number of rows, it's good practice to make sure the number of rows is capped.
+
+To set a limit, set the `maxRowsInTable` and `maxRowsInSubtable` properties in the constructor of your `RecordBuilder`.
+This can be hard-coded or it can come from configuration:
+
+```
+class MyRecordBuilder extends RecordBuilder
+{
+    public function __construct()
+    {
+        parent::__construct();
+        $this->maxRowsInTable = (int)Config::getInstance()->MyPlugin['datatable_archiving_maximum_rows'];
+        $this->maxRowsInSubtable = (int)Config::getInstance()->MyPlugin['datatable_archiving_maximum_rows_subtable'];
+
+        // we want to sort by the most important metric in our reports before we cut off rows
+        $this->columnToSortByBeforeTruncation = Metrics::INDEX_NB_VISITS;
+    }
+}
+```
+
+If you don't know what to use, you can set both values to `Config::getInstance()->General['datatable_archiving_maximum_rows_standard']`.
+
+Also note we set `columnToSortByBeforeTruncation` to make sure the rows with the least visits are the ones that get removed.
+
+Additionally, if your plugin provides metrics that should be aggregated together with an operation other than being `sum`-ed,
+you will need to set the `$columnAggregationOps` property:
+
+```
+class MyRecordBuilder extends RecordBuilder
+{
+    public function __construct()
+    {
+        parent::__construct();
+        
+        // ...
+
+        $this->columnAggregationOps = [
+            'my_max_metric' => 'max',
+            'my_min_metric' => 'min',
+            'my_other_metric' => function ($thisValue, $otherValue) {
```

Comment:
This is missing `$thisRow` and `$otherRow` arguments

Reply:
:+1: